src/java.base/share/classes/sun/net/www/http/KeepAliveCache.java

```diff
@@ -313,62 +316,48 @@ class ClientVector extends ArrayDeque<KeepAliveEntry> {
         this.nap = nap;
     }
 
+    /* return a still valid, idle HttpClient */
     HttpClient get() {
         lock();
         try {
             if (isEmpty()) {
                 return null;
             }
 
-            // Loop until we find a connection that has not timed out
-            HttpClient hc = null;
+            // check the most recent connection, use if still valid
             long currentTime = System.currentTimeMillis();
-            do {
-                KeepAliveEntry e = pop();
-                if ((currentTime - e.idleStartTime) > nap) {
-                    e.hc.closeServer();
-                } else {
-                    hc = e.hc;
-                    if (KeepAliveCache.logger.isLoggable(PlatformLogger.Level.FINEST)) {
-                        String msg = "cached HttpClient was idle for "
-                                + Long.toString(currentTime - e.idleStartTime);
-                        KeepAliveCache.logger.finest(msg);
-                    }
-                }
-            } while ((hc == null) && (!isEmpty()));
-            return hc;
-        } finally {
-            unlock();
-        }
-    }
-
-    /* return a still valid, unused HttpClient */
-    void put(HttpClient h) {
-        lock();
-        try {
-            if (size() >= KeepAliveCache.getMaxConnections()) {
-                h.closeServer(); // otherwise the connection remains in limbo
+            KeepAliveEntry e = peekFirst();
+            if ((currentTime - e.idleStartTime) > nap) {
```

Comment:
Since `peekFirst` can return null in theory, should we check for it? Currently `pop` is used which throws an exception instead, and probably means it doesn't happen in practice, but maybe it could be flagged in static source code analysis?